src/motor-operations.cc

```diff
@@ -84,19 +85,97 @@ static char test_acceleration_ok(float acceleration) {
+// has to be performed.
+class MotionQueueMotorOperations::StatusListener : public MotionQueue::EventReceiver {
+public:
+    StatusListener(MotionQueue *backend, MotorCoordSystem *current_position_steps)
+                        : backend_(backend), current_position_steps_(current_position_steps) {};
+    virtual void update_absolute_position();
+    void push_in_history(const LinearSegmentSteps &param, const int defining_axis_steps) {
+        struct history_slot *just_uploaded = execution_history.append();
+        for (int i = 0; i < MOTION_MOTOR_COUNT; ++i) {
+            just_uploaded->steps[i] = param.steps[i];
+        }
+        just_uploaded->max_axis_steps = defining_axis_steps;
+    }
+private:
+    MotionQueue *const backend_;
+
+    SlotHistory execution_history;
+    SlotPartial last_partial;
+    MotorCoordSystem *const current_position_steps_;
+};
+
+void MotionQueueMotorOperations::StatusListener::update_absolute_position() {
+    int i;
+    const struct QueueStatus status_register = backend_->get_status();
+    if (execution_history.size() == 0) return;
+    // pru_queue_size represent the real pru queue size, so we shrink the history queue until it's exactly the same as the
+    // pru one.
+    const unsigned pru_queue_size = (execution_history.back_index() + QUEUE_LEN - status_register.index) % QUEUE_LEN + 1;
+
+    while (execution_history.size() > pru_queue_size) {
+        for (i = 0; i < MOTION_MOTOR_COUNT; ++i) {
+            (*current_position_steps_)[i] += execution_history[0]->steps[i];
+        }
+        execution_history.pop_front();
+    }
+
+    //Now update the current executing slot
+    uint64_t delta;
+    for (i = 0; i < MOTION_MOTOR_COUNT; ++i) {
+        (*current_position_steps_)[i] -= last_partial[i];
+
+        delta = status_register.counter * abs(execution_history[0]->steps[i]);
+        delta /= LOOPS_PER_STEP;
+        delta /= execution_history[0]->max_axis_steps;
+
+        if ( execution_history[0]->steps[i] < 0) {
+            last_partial[i] = execution_history[0]->steps[i] + delta;
+        } else {
+            last_partial[i] = execution_history[0]->steps[i] - delta;
+        }
+
+        (*current_position_steps_)[i] += last_partial[i];
+    }
+}
+
+MotionQueueMotorOperations::MotionQueueMotorOperations(MotionQueue *backend) : backend_(backend) {
+    status_listener_ = new StatusListener(backend, &current_position_steps);
+    backend_->callbacks = status_listener_;
+}
+
+MotionQueueMotorOperations::~MotionQueueMotorOperations() { delete status_listener_; }
+
+const MotorCoordSystem MotionQueueMotorOperations::GetCurrentMotorPos() {
+    status_listener_->update_absolute_position();
+    return current_position_steps;
+}
+
 void MotionQueueMotorOperations::EnqueueInternal(const LinearSegmentSteps &param,
                                                  int defining_axis_steps) {
   struct MotionSegment new_element = {0};
   new_element.direction_bits = 0;
-
+  int i;
```

Comment:
What was the reason for this change ? You removed the 'int' from the for-loop declaration and put it here. Don't do that unless absolutely necessary to avoid programming errors.